linux/libdriver/ioctl/rollup.c

```diff
@@ -0,0 +1,96 @@
+#include <stdio.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <stdint.h>
+#include <string.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <errno.h>
+#include <linux/cartesi/rollup.h>
+#include "rollup.h"
+
+int rollup_driver_init(struct rollup_driver *me)
+{
+	int rc = 0;
+
+	me->fd = open("/dev/rollup", O_RDWR);
+	if (me->fd < 0)
+		return errno;
+
+	struct rollup_setup setup;
+	if ((rc = ioctl(me->fd, IOCTL_ROLLUP_SETUP, &setup))) {
+		rc = errno;
+		goto setup_failed;
+	}
+
+	me->tx->length = setup.tx.length;
+	me->tx->data = mmap((void *)setup.tx.data, setup.tx.length,
+	                    PROT_READ | PROT_WRITE,
+	                    MAP_SHARED | MAP_SHARED, me->fd, 0);
+	if (me->tx->data == MAP_FAILED) {
+		rc = ENOBUFS;
+		goto tx_map_failed;
+	}
+
+	me->rx->length = setup.rx.length;
+	me->rx->data = mmap((void *)setup.rx.data, setup.rx.length,
+	                    PROT_READ,
+	                    MAP_SHARED | MAP_SHARED, me->fd, 0);
+	if (me->rx->data == MAP_FAILED) {
+		rc = ENOBUFS;
+		goto rx_map_failed;
+	}
+	return 0;
+
+rx_map_failed:
+	munmap(me->tx->data, me->tx->length);
+tx_map_failed:
+setup_failed:
+	close(me->fd);
+	return rc;
+}
+
+void rollup_driver_fini(struct rollup_driver *me)
+{
+	munmap(me->tx->data, me->tx->length);
+	munmap(me->rx->data, me->rx->length);
+	close(me->fd);
+}
+
+int rollup_driver_write_output(struct rollup_driver *me, bool verifiable, uint64_t n)
+{
+	int type = verifiable
+		? IOCTL_ROLLUP_WRITE_VERIFIABLE_OUTPUT
+		: IOCTL_ROLLUP_WRITE_OUTPUT;
+
+	return ioctl(me->fd, type, (unsigned long) &n);
+}
+
+int rollup_driver_write_exception(struct rollup_driver *me, uint64_t n)
+{
+	return ioctl(me->fd, IOCTL_ROLLUP_WRITE_EXCEPTION, (unsigned long) &n);
+}
+
+int rollup_driver_loop(struct rollup_driver *me, uint8_t hash[32], uint64_t *n)
+{
+	memcpy(me->tx->data, hash, 32);
+	return ioctl(me->fd, IOCTL_ROLLUP_WAIT_FOR_INPUT, (unsigned long) n);
+}
+
+int rollup_driver_revert(struct rollup_driver *me)
+{
+	return ioctl(me->fd, IOCTL_ROLLUP_REJECT_INPUT);
+}
+
+void *rollup_driver_get_tx(struct rollup_driver *me, size_t *max)
+{
+	*max = me->tx->length;
```

Comment:
Better check if `max` is not `NULL`, and set it in this case. Same for the function below.